python/equistore-operations/equistore/operations/join.py

```diff
@@ -285,3 +301,117 @@ def join(
         return remove_dimension(tensor_joined, name="tensor", axis=axis)
     else:
         return tensor_joined
+
+
+def _intersection(tensors: List[TensorMap]):
+    """Create a new tensors list where keys are based on the intersection from all
+    tensors.
+
+    Blocks corresponding to keys that are not present in all tensor will be discarded.
+    """
+    # Constrauct a Labels object with intersected keys
+    all_keys = tensors[0].keys
+    for tensor in tensors[1:]:
+        all_keys = all_keys.intersection(tensor.keys)
+
+    # Create new blocks and discard bocks not present in all_keys
+    new_tensors = []
+    for tensor in tensors:
+        new_blocks = [block_to(tensor.block(key)) for key in all_keys]
```

Comment:
I don't know if `block_to` is doing the correct thing here. I introduce this because @bananenpampe had problems when trying to use join with a rascaline TensorMap.

Reply:
I don't think we want to do conversions here. If the user wants to join torch & numpy data, they need to call `equistore.to` themself.